src/module.js

```diff
@@ -0,0 +1,326 @@
+    opts = obj.merge({reloadDeps: true, resetEnv: true}, opts);
+    var toBeReloaded = [this];
+    if (opts.reloadDeps) toBeReloaded = this.dependents.concat(toBeReloaded);
+    await this.unload({forgetDeps: opts.reloadDeps, forgetEnv: opts.resetEnv});
+    await Promise.all(toBeReloaded.map(ea => ea.id !== this.id && ea.load()));
+    await this.load();
+  }
+  
+  // -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+  // change
+  // -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+  
+  async changeSourceAction(changeFunc) {
+    const source = await this.source();
+    const newSource = await changeFunc(source);
+    return this.changeSource(newSource, {evaluate: true});
+  }
+
+  async changeSource(newSource, options) {
+    const oldSource = await this.source();
+    return moduleSourceChange(this.System, this.id, oldSource, newSource, this.metadata.format, options);
+  }
+  
+  // -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+  // dependencies
+  // -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+  
+  get dependents() {
+    // which modules (module ids) are (in)directly import module with id
+    // Let's say you have
+    // module1: export var x = 23;
+    // module2: import {x} from "module1.js"; export var y = x + 1;
+    // module3: import {y} from "module2.js"; export var z = y + 1;
+    // `dependents` gives you an answer what modules are "stale" when you
+    // change module1 = module2 + module3
+    return graph.hull(graph.invert(computeRequireMap(this.System)), this.id)
+                .map(mid => module(this.System, mid));
+  }
+
+  get requirements() {
+    // which modules (module ids) are (in)directly required by module with id
+    // Let's say you have
+    // module1: export var x = 23;
+    // module2: import {x} from "module1.js"; export var y = x + 1;
+    // module3: import {y} from "module2.js"; export var z = y + 1;
+    // `module("./module3").requirements` will report ./module2 and ./module1
+    return graph.hull(computeRequireMap(this.System), this.id)
+                .map(mid => module(this.System, mid));
+  }
+  
+  // -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+  // module environment
+  // -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+
+  get env() {
+    const ext = this.System.get("@lively-env");
+    if (ext.loadedModules[this.id]) return ext.loadedModules[this.id];
+  
+    const env = {
+      loadError: undefined,
+      recorderName: "__lvVarRecorder",
+      dontTransform: [
+        "__lvVarRecorder",
+        "global", "self",
+        "_moduleExport", "_moduleImport",
+        "fetch" // doesn't like to be called as a method, i.e. __lvVarRecorder.fetch
+      ].concat(ast.query.knownGlobals),
+      recorder: Object.create(this.System.global, {
+        _moduleExport: {
+          get() {
+            return (name, val) => {
+              scheduleModuleExportsChange(this.System, this.id, name, val, true/*add export*/);
```

Comment:
Please be super careful about `this`! We are using here method syntax, not arrow functions, meaning that this isn't bound to the ModuleInterface object but the recorder.